[Bug]: corrosion breaks due to new `rustup toolchain list` format in rustup v1.28.0

### Current Behavior

Hi! @rami3l from the rustup team here.

After announcing [rustup v1.28.0 beta](https://internals.rust-lang.org/t/seeking-beta-testers-for-rustup-v1-28-0/22060), we have received a bug report that goes like so:

(After upgrading rustup from v1.27.1 to v1.28.0, corrosion is no longer able to find the toolchain it needs:)

> ```
> CMake Warning (dev) at build/_deps/corrosion-src/cmake/FindRust.cmake:363 (message):
>   Unexpected output from `rustc --version` for Toolchain
>   `stable-x86_64-unknown-linux-gnu`: ``.
> 
>   Ignoring this toolchain.
> Call Stack (most recent call first):
>   build/_deps/corrosion-src/cmake/Corrosion.cmake:63 (find_package)
>   build/_deps/corrosion-src/CMakeLists.txt:73 (include)
> This warning is for project developers.  Use -Wno-dev to suppress it.
> 
> -- Rust Toolchain: 
> CMake Error at build/_deps/corrosion-src/cmake/FindRust.cmake:23 (message):
>   Could not find toolchain ''
> 
>   Available toolchains:
> 
>     `nightly-x86_64-unknown-linux-gnu`
>     `nightly-2024-03-10-x86_64-unknown-linux-gnu`
>     `1.70-x86_64-unknown-linux-gnu`
>     `1.77-x86_64-unknown-linux-gnu`
> 
> Call Stack (most recent call first):
>   build/_deps/corrosion-src/cmake/FindRust.cmake:469 (_findrust_failed)
>   build/_deps/corrosion-src/cmake/Corrosion.cmake:63 (find_package)
>   build/_deps/corrosion-src/CMakeLists.txt:73 (include)
> ```
_https://github.com/rust-lang/rustup/issues/4127_

### Expected Behavior

The project builds correctly.

### Steps To Reproduce

Build any corrosion project with rustup v1.28.0.

The direct cause of this issue is the interaction between https://github.com/rust-lang/rustup/pull/3225 and the following section of corrosion:

https://github.com/corrosion-rs/corrosion/blob/b88ec09915a6b12d1aaec7e451ecf0e41a831030/cmake/FindRust.cmake#L337-L350

### Environment

```markdown
- OS: Fedora 41
```

(However, this error should be platform-independent.)

### Remarks

We can do a silent fix on our side, however, we would prefer informing you of this issue in advance so that we can figure out the best way forward together. I will be happy to work on a patch to corrosion if you'd like.

The original motivation of https://github.com/rust-lang/rustup/pull/3225 is that it might not be immediately clear to the end user what the active toolchain is, I wonder if you have the same issue in this project in that respect.

Also, we are interested in providing a machine-facing output mode. Please feel free to [provide suggestions](https://github.com/rust-lang/rustup/issues/3434) WRT how you'd expect this to be done.


Many thanks in advance, and happy holidays :)

--- a/cmake/FindRust.cmake
+++ b/cmake/FindRust.cmake
@@ -334,18 +334,18 @@ if (Rust_RESOLVE_RUSTUP_TOOLCHAINS)
     set(_DISCOVERED_TOOLCHAINS_VERSION "")
 
     foreach(_TOOLCHAIN_RAW ${_TOOLCHAINS_RAW})
-        if (_TOOLCHAIN_RAW MATCHES "([a-zA-Z0-9\\._\\-]+)[ \t\r\n]?(\\(default\\) \\(override\\)|\\(default\\)|\\(override\\))?[ \t\r\n]+(.+)")
+        if (_TOOLCHAIN_RAW MATCHES "([a-zA-Z0-9\\._\\-]+)[ \t\r\n]?(\\(active\\)|\\(active, default\\)|\\(default\\) \\(override\\)|\\(default\\)|\\(override\\))?[ \t\r\n]+(.+)")
             set(_TOOLCHAIN "${CMAKE_MATCH_1}")
             set(_TOOLCHAIN_TYPE "${CMAKE_MATCH_2}")
 
             set(_TOOLCHAIN_PATH "${CMAKE_MATCH_3}")
             set(_TOOLCHAIN_${_TOOLCHAIN}_PATH "${CMAKE_MATCH_3}")
 
-            if (_TOOLCHAIN_TYPE MATCHES ".*\\(default\\).*")
+            if (_TOOLCHAIN_TYPE MATCHES ".*\\((active, )?default\\).*")
                 set(_TOOLCHAIN_DEFAULT "${_TOOLCHAIN}")
             endif()
 
-            if (_TOOLCHAIN_TYPE MATCHES ".*\\(override\\).*")
+            if (_TOOLCHAIN_TYPE MATCHES ".*\\((active|override)\\).*")
                 set(_TOOLCHAIN_OVERRIDE "${_TOOLCHAIN}")
             endif()
 

